content/posts/side-effects.md

```diff
@@ -0,0 +1,113 @@
+
+```c#
+public static async Task CreateBlob(this IHandlerContext context, string blobPrefix)
+{
+    var unique = Guid.NewGuid();
+    var url = blobPrefix + "-" + unique.ToString();
+    var sideEffectRecord = new BlobSideEffectRecord {
+        Attempt = context.Attempt,
+        IncomingMessageId = context.IncomingMessageId,
+        Url = url;
+    };
+
+    await context.TransactionalStore.Add(sideEffectRecord);
+    await context.BlobClient.UploadBlob(url);
+}
+```
+
+First, in line 3 we generate our unique value. Then in line 5-9 we construct the side effect record. As we mentioned previously, there is nothing to do when publishing blobs so this record is only needed for cleanup. Notice that in line 11 we store the side effect before we actually upload the blob in line 12. This ensures that we won't end up in a situation with an uploaded blob both without the corresponding side effect record. The cleanup loops through all side effects recorded for a given incoming message and cleans up the ones that have `Attempt` value different than the ID of the successful attempt.
+
+Last but not least, conditions number one and two. How do we ensure the side effects are indeed published (and the ones that are to be cleaned up are removed) after we know it has been processed successfully but before it is consumed. This one is relatively simple. We need to inject the publishing of the side effects right after the transaction store transaction is committed. This ensures that we don't generate _ghost side effects_ -- side effects that carry the state that has not yet been made durable.
+
+We also need to make sure the publishing happens before we clean up the transaction information from the store.
+
+### Summary
+
+Let's recap what the conditions for a correct side effects algorithm are:
+ - we want all side effects to be visible before the incoming message is consumed
+ - we don't want any side effects to be visible if the incoming message have not been processed
+ - side effects that are published all come from the same processing attempt
+ - by the time the message is processed successfully (consumed), all side effects that resulted from non-successful attempts are cleaned up
+
+Our sketched up algorithm satisfies all four. The side effects are made visible when the transactional store transaction completes. All side effects come from the same (successful) attempt because we ensure their identity contains a unique value. We make sure to leave no garbage by recording the identity of side effects before thy are created and cleaning up data generated by failed attempts.
```

Comment:
```suggestion
Our sketched up algorithm satisfies all four. The side effects are made visible when the transactional store transaction completes. All side effects come from the same (successful) attempt because we ensure their identity contains a unique value. We make sure to leave no garbage by recording the identity of side effects before they are created and cleaning up data generated by failed attempts.
```